DigitalLearningSolutions.Data.Tests/Services/SessionServiceTests.cs

```diff
@@ -0,0 +1,166 @@
+﻿namespace DigitalLearningSolutions.Data.Tests.Services
+{
+    using System;
+    using System.Linq;
+    using System.Transactions;
+    using DigitalLearningSolutions.Data.Models;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Data.Tests.Helpers;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    internal class SessionServiceTests
+    {
+        private SessionService sessionService;
+        private SessionTestHelper sessionTestHelper;
+
+        [SetUp]
+        public void Setup()
+        {
+            var connection = ServiceTestHelper.GetDatabaseConnection();
+            sessionService = new SessionService(connection);
+            sessionTestHelper = new SessionTestHelper(connection);
+        }
+
+        [Test]
+        public void StartOrRestartSession_Should_Start_Users_First_Session()
+        {
+            using (new TransactionScope())
+            {
+                // When
+                const int candidateId = 29;
+                const int customisationId = 100;
+                var sessionId = sessionService.StartOrRestartSession(candidateId, customisationId);
+
+                // Then
+                var newSession = sessionTestHelper.GetSession(sessionId);
+
+                newSession.Should().NotBeNull();
+                SessionTestHelper.SessionsShouldBeApproximatelyEquivalent(
+                    newSession!,
+                    new Session(
+                        sessionId,
+                        candidateId,
+                        customisationId,
+                        DateTime.Now,
+                        0,
+                        true
+                    )
+                );
+            }
+        }
+
+        [Test]
+        public void StartOrRestartSession_Should_Start_Subsequent_Session()
+        {
+            using (new TransactionScope())
+            {
+                // When
+                const int candidateId = 101;
+                const int customisationId = 1240;
+                var sessionId = sessionService.StartOrRestartSession(candidateId, customisationId);
+
+                // Then
+                var newSession = sessionTestHelper.GetSession(sessionId);
+
+                newSession.Should().NotBeNull();
+                SessionTestHelper.SessionsShouldBeApproximatelyEquivalent(
+                    newSession!,
+                    new Session(

```

Comment:
Could have a helper method to create a session, like we have for e.g. course content, so that you can only override the fields you care about 🙂

Reply:
I'm not quite sure what you mean here. These lines are to check the values in the session that has just been inserted are correct using equality for most of the fields, but using approximate equality for the `LoginTime` (because I won't know precisely which millisecond the event will be created at). How could I use a create-session method to help with this? 🙂

Reply:
Sorry, I just meant to create the Session object. We already have a few methods like that e.g. to create a CourseContent object. So instead of having to call the constructor with all the fields that requires, you call your helper method which has defaults for every constructor parameter so you only need to specify the parameters you really care about in your test. Does that make sense? 🙂

Reply:
Ah I see! I've done that in a4825ca 🙂